rclpy/src/rclpy/action_client.cpp

```diff
@@ -0,0 +1,321 @@
+// Copyright 2021 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+
+#include <memory>
+#include <string>
+
+#include "rclpy_common/common.h"
+#include "rclpy_common/exceptions.hpp"
+#include "rclpy_common/handle.h"
+
+#include "action_client.hpp"
+#include "utils.hpp"
+
+namespace rclpy
+{
+
+void
+ActionClient::destroy()
+{
+  rcl_action_client_.reset();
+  node_.destroy();
+}
+
+ActionClient::ActionClient(
+  Node & node,
+  py::object pyaction_type,
+  const char * action_name,
+  const rmw_qos_profile_t & goal_service_qos,
+  const rmw_qos_profile_t & result_service_qos,
+  const rmw_qos_profile_t & cancel_service_qos,
+  const rmw_qos_profile_t & feedback_topic_qos,
+  const rmw_qos_profile_t & status_topic_qos)
+: node_(node)
+{
+  rosidl_action_type_support_t * ts =
+    static_cast<rosidl_action_type_support_t *>(rclpy_common_get_type_support(
+      pyaction_type.ptr()));
+  if (!ts) {
+    throw py::error_already_set();
+  }
+
+  rcl_action_client_options_t action_client_ops = rcl_action_client_get_default_options();
+
+  action_client_ops.goal_service_qos = goal_service_qos;
+  action_client_ops.result_service_qos = result_service_qos;
+  action_client_ops.cancel_service_qos = cancel_service_qos;
+  action_client_ops.feedback_topic_qos = feedback_topic_qos;
+  action_client_ops.status_topic_qos = status_topic_qos;
+
+  rcl_action_client_ = std::shared_ptr<rcl_action_client_t>(
+    new rcl_action_client_t,
+    [node](rcl_action_client_t * action_client)
+    {
+      // Intentionally capture node by value so shared_ptr can be transferred to copies
+      rcl_ret_t ret = rcl_action_client_fini(action_client, node.rcl_ptr());
+      if (RCL_RET_OK != ret) {
+        // Warning should use line number of the current stack frame
+        int stack_level = 1;
+        PyErr_WarnFormat(
+          PyExc_RuntimeWarning, stack_level, "Failed to fini publisher: %s",
+          rcl_get_error_string().str);
+        rcl_reset_error();
+      }
+      delete action_client;
```

Comment:
@sloretz meta: in retrospect, I wonder if we should've used a Python allocator-backed `std::allocator` for this.

Reply:
Oh yeah! You mean like the allocator `rclpy::Client` is using? Looks lt's the only one using it.

https://github.com/ros2/rclpy/blob/fbe9ad12b806345a3b3c0578af7e21d1d6bc7bea/rclpy/src/rclpy/client.cpp#L55-L70

Reply:
Indeed.